### DISCUSSION TITLE
How to use multiple domains in a load-balanced environment

### DISCUSSION THREAD

**Discussion by nichols-green** [directus/directus#8432] [Archive]
When used in a load balanced situation with two nodes and a load balancer, the `public_url` env var breaks the load balance config. 

server 1 is addressed as `mysite-1.mydomain.org`
server 2 is addressed as `mysite-2.mydomain.org`
The LB is addressed as `mysite.mydomain.org`.

In the Directus config, I've set the `public_url` to `mysite.mydomain.org`. This works for the customer facing site and the email invitations. This setup fails when attempting to reach the `-1`, `-2` sites directly. This is essential as the load balancer health check attempts to reach the downstream servers by their `-1`, `-2` addresses. 

Essentially we can run the site without health checks and everything works, or we can enable health checks as is our policy and the site breaks. The load balancer isn't able to get a response from the health check endpoints, and therefore has no active downstream servers to route to. 

Lastly, I set the `public_url` variable to `/` and everything worked as expected. The load balancer health checks worked, and the site was reachable at all three addresses. I thought this was the solution, until I tried to invite a member of my team to the site and the links in the email were all just relative links prefixed with `/`

I would suggest separating thi `public_url` variable into `public_url` and `host_url`, or something along those lines to allow the use of `/` as the site host and still have `mysite.mydomain.org` in the email prefix and other areas where that would be used. 

### To Reproduce

Deploy a site to two or more servers, with a loadbalancer. Assign the servers addresses, and assign the loadbalancer an address. Set the `public_url` env var to the value of the load balancer address. Ensure health checks are off. The site should work as expected, but the two direct addresses will not work. 

Set the `public_url` value to `/`, and the site should be working as expected. Invite someone to the site, and check the email links. The links will not be complete.

**Comment by rijkvanzanten** [comment]
> In the Directus config, I've set the public_url to mysite.mydomain.org. This works for the customer facing site and the email invitations. This setup fails when attempting to reach the -1, -2 sites directly. This is essential as the load balancer health check attempts to reach the downstream servers by their -1, -2 addresses.

This shouldn't necessarily matter in Directus itself. That public URL is only used for things like password-reset emails, and other user-facing names. That being said, it doesn't affect the /server/health endpoint at all (Directus itself runs fine regardless of what value you use for PUBLIC_URL). I'm a little confused as to why that would affect the health check at all. How are you health checking currently?

**Reply by nichols-green** [reply]
Our IT team expects a specific response, so we have a custom endpoint that calls the ServerService.Health() and sends the expected response depending on the status result. 

It is odd to me that the site can't be reached at the `-1`, `-2` addresses when the variable is set to `mysite.mydomain.org`.

**Reply by nichols-green** [reply]
I should clarify that the custom health endpoint, and the site itself are unreachable at the `-1`, and `-2` addresses when the public_url is set to `mysite.mydomain.org`

**Reply by nichols-green** [reply]
And just to be thorough, I tried the built-in health endpoint, and got `The service is unavailable` as a response. Attempting to load the root app leaves me with a blank page, and connection reset errors in console.

**Reply by rijkvanzanten** [reply]
When you say

> It is odd to me that the site can't be reached

do you mean the Directus app?

**Reply by nichols-green** [reply]
Yes. Though that's probably not accurate. The site is resolving correctly, but not rendering. It is attempting to load the bundles, but erroring.

**Reply by nichols-green** [reply]
![image](https://user-images.githubusercontent.com/82182892/135313218-dbea84bd-12a1-492f-9d93-7699d104959b.png)

**Reply by rijkvanzanten** [reply]
Yes, that part checks out, as it relies on the public url to prefix the "base" meta tag in html to find the static assets (to support for any deeply nested route path). That's something we can resolve with your suggestion of adding a dedicated "host" vs "public" env var. However, that doesn't explain the failing server health custom endpoint, as that shouldn't be affected by any of this 🤔

**Reply by nichols-green** [reply]
Here's the full code of that extension:

```
module.exports = function registerEndpoint(router, { services, exceptions }) {
	const { ServerService } = services;
	const { ServiceUnavailableException } = exceptions;

	router.get('/', (req, res, next) => {
		const ThisServerService = new ServerService({ schema: req.schema, accountability: req.accountability });

        ThisServerService
	        .health()
	        .then((results) => {
                        if (results.status === 'ok') {
                              return res.send('UP-100');
                        } else {
                              return res.status(503).send("DOWN-200");
                        }
                })
	        .catch((error) => {
		        return next(new ServiceUnavailableException(error.message));
	        });
	});
};
```

**Reply by nichols-green** [reply]
@rijkvanzanten Any idea on a timeline for the env var additions/changes?

**Reply by rijkvanzanten** [reply]
That endpoint logic looks good to me, and from what I can tell isn't affected at all by the PUBLIC_URL environment variable 🤔 

With custom endpoint:

```js
module.exports = function (router) {
	router.get('/', (req, res) => res.send('Example'));
};
```

When I set my PUBLIC_URL env var to something wrong (like `PUBLIC_URL="http://rijks.website"`), I'm still able to access my custom endpoint straight through the original location IP:

<img width="567" alt="CleanShot 2021-09-30 at 11 44 22@2x" src="https://user-images.githubusercontent.com/9141017/135487903-8df82874-7e42-417f-8190-0fa6eeec7ea9.png">

Are we sure this isn't caused by some other network configuration or environment change besides the public url flag?

> Any idea on a timeline for the env var additions/changes?

This is currently an open feature request. If enough people upvote it we'll look into implementing it if and when possible. As of now, the only way to get an accurate timeline or release date is to [sponsor](https://docs.directus.io/getting-started/support.html#sponsored-work) this task. We recognize that this may be an important feature/fix, but we are a small open-source organization with a lot to triage and complete. If you can't sponsor this ticket, then the next best thing is to increase its priority by giving it a 👍 . Or contribute a pull-request... after all, this is OSS!

**Reply by nichols-green** [reply]
> Are we sure this isn't caused by some other network configuration or environment change besides the public url flag?

I was asking this question myself and so I tested switching the value of the public_url variable from the site address to `/` and back. The custom endpoint returns `service unavailable` when the public_url is set to anything other than `/` or the address I am attempting to reach. I get the same result when attempting to reach the `/server/health` endpoint. 

> This is currently an open feature request. If enough people upvote it we'll look into implementing it if and when possible. As of now, the only way to get an accurate timeline or release date is to sponsor this task. We recognize that this may be an important feature/fix, but we are a small open-source organization with a lot to triage and complete. If you can't sponsor this ticket, then the next best thing is to increase its priority by giving it a 👍 . Or contribute a pull-request... after all, this is OSS!

Happy to contribute! This one feels a little beyond me, but I'll give it a shot. After all, the last one felt a little beyond me as well and that worked out beautifully.

**Reply by nichols-green** [reply]
Also, your example controller doesn't make use of ServerService, and I suspect it is within this service that there is a dependency on public_url, either directly, or while loading other deps.

**Reply by rijkvanzanten** [reply]
> Happy to contribute! This one feels a little beyond me, but I'll give it a shot. After all, the last one felt a little beyond me as well and that worked out beautifully.

I think it could be doable by adding a new "HOST_URL" env var here https://github.com/directus/directus/blob/e95404a1daea19b75f61a88ddc877b1f82ef5cb4/api/src/env.ts#L19 that defaults to "PUBLIC_URL" (so we don't require the user to set both if they don't need it), and then effectively find and replace all uses of PUBLIC_URL with the new HOST_URL (except for the one that creates the links in the emails) 🤔

**Reply by nichols-green** [reply]
OK I was wrong. The custom endpoints/health endpoint failure was due to a misconfigured uploads directory. So this failure is only related to the loading of dependencies. Should be a pretty quick fix.

**Reply by rijkvanzanten** [reply]
Ahh yeah that explains!

**Comment by nichols-green** [comment]
This was solved somewhere between RC.76 and RC.96 and is no longer a concern.
